examples/simon.py

```diff
@@ -0,0 +1,339 @@
+'''
+
+import cirq
+import random
+import numpy as np
+import copy
+import sympy
+import itertools
+
+# Qubit preparation
+
+number_qubits = #Number of qubits
+
+def main(number_qubits):
+
+    circuit_sampling = number_qubits-1
+
+    #Create the qubits which are used within the circuit
+
+    first_qubits = [cirq.GridQubit(i, 0) for i in range(number_qubits)]
+    second_qubits = [cirq.GridQubit(i, 0) for i in range(number_qubits, 2*number_qubits)]
+
+    the_activator = cirq.GridQubit(2*number_qubits, 0)
+
+    #Create the qubits that can be used for large-input Toffoli gates
+    ancilla = []
+    for v in range(2*number_qubits+1, 3*number_qubits):
+        ancilla.append(cirq.GridQubit(v, 0))
+
+    #Create the function that is inputted into the algorithm (secret!)
+
+    domain = []
+    selector = []
+    co_domain = []
+    fixed = []
+
+    for k in range(0, 2**number_qubits):
+        domain.append(k)
+        selector.append(k)
+        co_domain.append(False)
+        fixed.append(k)
+
+    #Create the "secret string"
+    s = domain[random.randint(0, len(domain)-1)]
+
+    #Create the "secret function"
+    for g in range(0, int((2**number_qubits)/2)):
+        v = random.choice(selector)
+        x = random.choice(domain)
+        co_domain[x] = v
+        co_domain[x^s] = v
+        del selector[selector.index(v)]
+        del domain[domain.index(x)]
+        if (s != 0):
+            del domain[domain.index(x^s)]
+
+    secret_function = [fixed, co_domain]
+
+    oracle = make_oracle(ancilla, secret_function, first_qubits, second_qubits, s, the_activator)
+
+    c = make_simon_circuit(first_qubits, second_qubits, oracle)
+
+    #Sampling the circuit
+
+    simulator = cirq.Simulator()
+    result = simulator.run(c, repetitions=number_qubits-1)
+    final = result.histogram(key='y')
+    print("Secret String: "+str(s))
+    print("Secret Function (Domain and Co-Domain): "+str(secret_function))
+    final = str(result)[str(result).index("y")+2:len(str(result))].split(", ")
+    last = []
+    for i in range(0, number_qubits-1):
+        holder = []
+        for j in final:
+            holder.append(int(j[i]))
+        holder.append(0)
+        last.append(holder)
+
+    print("Results: "+str(last))
+    return [last, secret_function, s, last]
+
+
+
+
+def make_oracle(ancilla, secret_function, first_qubits, second_qubits, s, the_activator):
+
+    #Hard-code oracle on a case-by-case basis
+
+
+    for o in range(0, len(secret_function[0])):
+        counter = 0
+        for j in list(str(format(secret_function[0][o], "0"+str(number_qubits)+"b"))):
+            if (int(j) == 0):
+                yield cirq.X.on(first_qubits[counter])
+            counter = counter+1
+        yield apply_n_qubit_tof(ancilla, first_qubits+[the_activator])
+        counter = 0
+        for j in list(str(format(secret_function[0][o], "0"+str(number_qubits)+"b"))):
```

Comment:
This method has a lot of repeated code. Extract the character counting loop into a method. Only initialize the strings you're making for 0 and for 1 a single time. Use `cirq.X.on_each(first_qubits[:count])` instead of mixing the counting with the gate making.